pkg/mcp/server/server.go

```diff
@@ -232,12 +249,16 @@ func (con *connection) receive() {
 	for {
 		req, err := con.stream.Recv()
 		if err != nil {
-			if status.Code(err) == codes.Canceled || err == io.EOF {
+			code := status.Code(err)
+			if code == codes.Canceled || err == io.EOF {
 				scope.Infof("MCP: connection %v: TERMINATED %q", con, err)
 				return
 			}
+			ctx, err := tag.New(context.Background(),
+				tag.Insert(ErrorTag, err.Error()),
+				tag.Insert(ErrorCodeTag, strconv.FormatUint(uint64(code), 10)))
+			stats.Record(ctx, RecvFailures.M(1))
```

Comment:
nit: might be worth moving to a `recordError(err error, code int64, stat measure)` method.